Swarm refactored

elona-dup opened the pull request.

elona-dup committed:
Added the notebook with the particle swarn optimization

elona-dup committed:
Added Particle and Swarm class

elona-dup committed:
Modified the init class for swarm particle opt

nesrnesr committed:
Constrain scheduler config to what's configurable (not constant)

Change interface of Swarm to define the environment constrains to run the experiments in.

nesrnesr committed:
Added flake8 file config

nesrnesr committed:
Swarm is able to draw stats instead of __init__

Fixed visualizer plotting figure
Server is now printable

nesrnesr committed:
Changed SchedulerConfig attributes which are returned in a list

Changed job generation parameters
Improved boundary checks on Particles

nesrnesr committed:
Changed Scheduler's logic to Reconfigure and Power off when possible

Redefined poweroffs and reconfiguration in a Job.

elona-dup committed:
Updated how jobs are generated using the article from Prof DaCosta

elona-dup committed:
Updated how particles behave at bounds using reflection

elona-dup committed:
Updated conditions for decision making based on parameters

elona-dup committed:
Added the ability to keep some stats of each epoch

elona-dup committed:
Added the ability to get a convergence graph from the swarm

elona-dup committed:
Dataclass to keep track of the stats at each epoch

elona-dup committed:
Added the possibility to activate/deactivate the different components of the scheduler to enable experimentation

elona-dup committed:
Cleaned up unecessary comments

elona-dup committed:
Added the ability to save the stats and the best configs to csv at the end of the runs

elona-dup committed:
Updated the seeding to ensure fair testing

elona-dup committed:
Updated the graph to only draw the mean cost

elona-dup committed:
Added the different set up for each experiments to be conducted in one run

elona-dup committed:
Adjusted the number of jobs used in each experiment to get more meaningful results

elona-dup committed:
added the number of server count in the condition for to reconfigure when using parameters

elona-dup committed:
Added all the different experiments performed

nesrnesr commented on Naive greedy alg/scheduling/Experiments.py:
```diff
@@ -89,3 +59,72 @@ def _generate_job(self, job_count, server_count, num):
         )
+
+    def _get_makespan(self, mass, disparity):
+        mu = numpy.log(mass / disparity)
```
Prefer using math over numpy
https://devdocs.io/python~3.7/library/math#math.log
https://devdocs.io/python~3.7/library/math#math.sqrt

elona-dup replied:
I don't think it makes much difference as we need to import numpy for the seeding of scipy distribution, but ok.

nesrnesr replied:
It is preferable to use the most appropriate tools for our needs. It would be restraining. If you use NumPy for everything then you won't be easily able to remove it if you decide not to to use anymore. It doesn't answer a specific need.

nesrnesr commented on Naive greedy alg/scheduling/Scheduler.py:
```diff
@@ -31,7 +32,9 @@ def random(cls):
         c.reconfig_prob = uniform(0.001, 1.0)
         c.reconfig_weight = uniform(0.01, 1.0)
-        c.shutdown_time = uniform(370, 1200)
+        c.shutdown_time_1 = uniform(370, 1200)
```
Naming needs to be more explicit

nesrnesr commented on Naive greedy alg/scheduling/Scheduler.py:
```diff
@@ -133,16 +136,20 @@ def update_schedule(self, time):
-            )
-            self._start_job(power_off)
-            av_servers.remove(server)
+            if (
```
For readability, the condition could be wrapped into a function like shutdown_enabled, shutdown_possible.

nesrnesr commented on Naive greedy alg/scheduling/Swarm.py:
```diff
@@ -23,14 +24,19 @@ def __init__(self, seed_num, num_particles, num_srvs, num_exp=10):
+    def run_epochs(self, num_epochs, draw_stats=False, draw_graph=False):
+        epochs_stats = []
         for i in range(num_epochs):
             print("epoch: ", i)
```
Remove print

elona-dup replied:
I have found the prints pretty useful to monitor how quickly the particles converge and to then decide on the number of epochs to use. I will comment them, if you think that they must be removed, then you can remove them.

nesrnesr replied:
You can use the Logger instead of the prints, and change its level to enable/disable the prints

nesrnesr commented on Naive greedy alg/scheduling/Swarm.py:
```diff
@@ -66,6 +80,12 @@ def _draw_stats(self, num_epoch, particle_idx, stats, config):
         print("Tuning configs:\n", config)
 
+    def _draw_graph(self, num_epoch, stats):
+        visualizer = Visualizer()
```
The visualizer should not be part of the Swarm. Return instead your list of stats from run_epochs and visualize the result from the main.

nesrnesr commented on Naive greedy alg/scheduling/Scheduler.py:
```diff
@@ -74,14 +63,23 @@ def to_dict(self):
+    ):
         self.servers = [Server(i) for i in range(server_count)]
         self.conf = conf
+        self.reconfig_enable = reconfig_enable
```
It should be _enabled.

nesrnesr commented on Naive greedy alg/scheduling/Scheduler.py:
```diff
@@ -123,33 +114,47 @@ def update_schedule(self, time):
+        if self.power_off_enable:
+            for server in list(av_servers):
+                if self.param_enable:
+                    if not self._shutdown_server(av_servers):
```
This should be moved up above if self.param_enable, it might be the cause of the infinite loop.

nesrnesr commented on Naive greedy alg/scheduling/Scheduler.py:
```diff
@@ -123,33 +114,47 @@ def update_schedule(self, time):
+        # Turn off servers
+        if self.power_off_enable:
+            for server in list(av_servers):
+                if self.param_enable:
+                    if not self._shutdown_server(av_servers):
+                        break
+                    if (
+                        0.5
+                        < (
+                            (len(av_servers) / len(self.servers))
+                            ** self.conf.shutdown_weight
+                        )
+                        * self.conf.shutdown_prob
+                    ):
+                        if random() < self.conf.shutdown_time_prob:
+                            shutdown_duration = self.conf.shutdown_time_1
+                        else:
+                            shutdown_duration = self.conf.shutdown_time_2
```
We could move this in a separate function and merge the shutdown task creation code
The function could return a boolean (indicating if the server can be shut down) and the appropriate shutdown_duration
This will allow having a clearer workflow.

Smth along these lines:

        # Turn off servers
        if self.power_off_enable:
            for server in list(av_servers):
                if not self._shutdown_server(av_servers):
                    break

                shutdown, duration = self._allow_shutdown(av_servers)
                if shutdown:
                    power_off = Job.make_power_off(
                        [server], start_time=time, duration=duration
                    )
                    self._start_job(power_off)
                    av_servers.remove(server)

nesrnesr commented on Naive greedy alg/scheduling/Swarm.py:
```diff
@@ -33,6 +33,12 @@ def run_epochs(self, num_epochs, draw_stats=False, draw_graph=False):
             self._draw_graph(
                 num_epochs, pd.DataFrame([stat.to_dict() for stat in epochs_stats])
             )
+        #   save final best config
```
This logic does not belong to the Swarm.
Applying the previous comment regarding the visualizer, you can move this logic in the main

nesrnesr commented on Naive greedy alg/scheduling/__init__.py:
```diff
@@ -1,14 +1,63 @@
+from .Experiments import Experiments
+from .Scheduler import SchedulerConfig
+
+# python -m scheduling
```
To remove (same for the swarm)

nesrnesr commented on Naive greedy alg/scheduling/__init__.py:
```diff
@@ -1,14 +1,63 @@
+    n_servers = 10
+    n_expts = 30
+    seed = 2
+    # run expt with no reconfig and no power off
```
I suggest we do 1 function that takes the csv name and **kwargs and then unpack the dictionary when calling run_exps, like run_expts(**kwargs)

elona-dup replied:
Seems fine to me. You can implement it if you have time

nesrnesr commented on Naive greedy alg/scheduling/Scheduler.py:
```diff
@@ -195,7 +195,7 @@ def _is_job_reconfigurable(self, job, av_servers, time):
 
         extra_srv_count = min(job.max_server_count - job.server_count, len(av_servers))
-        if self.param_enable:
+        if extra_srv_count > 0 and self.param_enable:
```
The added condition should not be necessary as the function is called only if there are available servers and as is_reconfigurable already checks that there is room for at least one more server.

nesrnesr commented:
I advise you to first resolve the conflicts

nesrnesr committed:
Use math instead of numpy when possible

nesrnesr committed:
Improves variable naming

Rename shut_down_time_1 and 2 to shut_down_time_short and long

nesrnesr committed:
Encasulates sets of experiments.

nesrnesr committed:
Uses structLog in top of logging

nesrnesr committed:
Put specific experiments in a seperate file

nesrnesr committed:
Rework the Swarm interface

Transform EpochStats into EpochCost
Encapsulate csv creation in the Visualizer
Move drawing logic outside of the Swarm

nesrnesr committed:
Added a new experiment to ExpermientTest
